tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11475,56 +11557,68 @@ void hlsl::DiagnoseTranslationUnit(clang::Sema *self) {
     }
   }
 
-  if (self->getLangOpts().IsHLSLLibrary) {
-    ValidatePatchConstantFunctionsExist(self);
-    ValidateNoRecursionInTranslationUnit(self);
-    return;
-  }
+  // Now check for recursion, and check for patch constant function
+  // reachabililty Validation methods differ depending on whether this is a
+  // library shader or not.
 
   // TODO: make these error 'real' errors rather than on-the-fly things
   // Validate that the entry point is available.
   DiagnosticsEngine &Diags = self->getDiagnostics();
   FunctionDecl *pEntryPointDecl = nullptr;
   FunctionDecl *pPatchFnDecl = nullptr;
```

Comment:
pPatchFnDecl should be moved into the for (FunctionDecl *FDecl : FDeclsToCheck) { loop.